hack/build_channel_catalog.sh

```diff
@@ -58,48 +64,13 @@ else
     bundle_versions_repo_url="https://$BUNDLE_VERSIONS_REPO"
 fi
 
-log "Cloning $BRANCH_CHANNEL branch from $BUNDLE_VERSIONS_REPO into $saas_operator_dir_base"
-git clone --branch "$BRANCH_CHANNEL" "$bundle_versions_repo_url" "$saas_operator_dir_base"
+log "Cloning $BUNDLE_VERSIONS_REPO into $saas_operator_dir_base"
+git clone "$bundle_versions_repo_url" "$saas_operator_dir_base"
 
-prev_operator_version=""
-removed_versions=""
+# if the line contains SKIP it will not be included
 if [[ -s "$bundle_versions_file" ]]; then
     log "$bundle_versions_file exists. We'll use to determine current version"
-    if [[ "$REMOVE_UNDEPLOYED" == "true" ]]; then
-        log "Checking if we have to remove any versions more recent than deployed hash"
-        # this will need to be modified if at any point we move to a canary like deployment
-        # where not all environments are deployed with the same operator version
-        deployed_hash=$(
-            curl -s -H "Authorization: Basic $(echo -n $APP_INTERFACE_USERNAME:$APP_INTERFACE_PASSWORD | base64)" \
-                -g "https://$APP_INTERFACE_BASE_URL/graphql?query={saas_files:saas_files_v1{name,resourceTemplates{name,targets{namespace{environment{name,labels}},ref}}}}" | \
-                jq -r '.data.saas_files[] | select(.name=="saas-'$OPERATOR_NAME'") | .resourceTemplates[].targets[] | select(.namespace.environment.labels | contains("\"type\":\"production\"")) | .ref' | \
-                uniq
-        )
-
-        log "Current deployed hash is $deployed_hash"
-
-        new_bundle_versions_file=$(mktemp -p "$temp_dir" new_bundle_versions_file.XXXX)
-        delete=false
-        # Sort based on commit number
-        for version in $(sort -t . -k 3 -g "$bundle_versions_file"); do
-            if [[ "$delete" == false ]]; then
-                echo "$version" >> "$new_bundle_versions_file"
-
-                short_hash=$(echo "$version" | cut -d- -f2)
-
-                if [[ "$deployed_hash" == "${short_hash}"* ]]; then
-                    log "found deployed hash in the bundle versions repository"
-                    delete=true
-                fi
-            else
-                log "Adding $version to removed versions"
-                removed_versions="$version $removed_versions"
-            fi
-        done
-        [[ -n "$removed_versions" ]] && log "The following versions will be removed: $removed_versions"
-        cp "$new_bundle_versions_file" "$bundle_versions_file"
-    fi
-    prev_operator_version=$(tail -n 1 "$bundle_versions_file")
+    prev_operator_version=$(grep -v SKIP "$bundle_versions_file" | tail -n 1)
```

Comment:
From what I can tell, this will only allow us to `SKIP` one or more of the *latest* versions. In other words, editing my bundle versions file from

```
v1.0.0
v1.0.1
v1.0.2
v1.0.3
```
to
```
v1.0.0
v1.0.1
v1.0.2 SKIP
v1.0.3 SKIP
```
will cause my build of `v1.0.4` to use `replaces: v1.0.1`. Cool.

But if I do something like this:
```
v1.0.0
v1.0.1
v1.0.2 SKIP
v1.0.3
```
it'll be a no-op. Specifically, it will *not* cause the index to be rebuilt such that `v1.0.3` replaces `v1.0.1`.

I recognize that this would be a tough one to solve; and maybe we don't want to do it right away. But we should at least leave a detailed explanation in the README and/or in a comment here.

Reply:
So you want to skip a version after a newer one has already been deployed... That's not going to work, indeed... The current system will work as expected if we SKIP every version after the wrong one

Reply:
Actually, I don't think it'll matter except in the case of trying to rebuild a catalog from scratch. In the above example, bundle 1.0.3 will be built with replaces 1.0.1 because when the 1.0.3 bundle was created the history looked like:

```shell
v1.0.0
v1.0.1
v1.0.2 SKIP
```

When bundle 1.0.4 is created, the history will look like this:

```shell
v1.0.0
v1.0.1
v1.0.2 SKIP
v1.0.3
```

So the last version will be selected at v1.0.3 for the 1.0.4 bundle, which is correct.  The 1.0.3 bundle is already in the catalog with the correct replaces (v1.0.1) and we aren't recreating that bundle.

If we really need this exact history though, I think this can be accomplished by changing the logic slightly.  Create a new set of versions from the existing one, taking out each entry that has SKIP next to it.  It should basically be a list of versions created like:

grep -v SKIP "$bundle_versions_file" 

That would get us a valid upgrade chain accepting any number of SKIP versions.  However, since we are only ever creating the latest bundle except in an emergency I'm not sure where this logic would even be used in this script.

Reply:
Rewriting the skips list in a CSV is basically rewriting history.  I don't see how it can be done with automation.  The bundles have to be regenerated with the same bundle version/sha combination to avoid breaking the upgrade change.  This could be done manually of course, but then the git repo would not reflect what is actually in a bundle.  To have the repo reflect the corrected bundle images you would probably have to re-write git history for those versions we need to skip.  Then you have to address the old bundle image already existing on the openshift nodes or else the corrected bundle images won't be pulled.  If that isn't done as well then the upgrade chain is not going to be tested in stage as it would be implemented in prod.  IMO, this is worse than the heavy-handed approach @rporres suggested.

I'm not convinced this issue is a major concern we need to plan to address.  IMO, @rporres has a means to handle it should it occur.  The recovery process ensures that the upgrade chain will be tested in stage before being promoted to pod.

Reply:
> Rewriting the skips list in a CSV is basically rewriting history. I don't see how it can be done with automation.

I don't think this is right. We are dropping a fresh CSV every time we run this script. That's the thing we're computing `replaces` for. What we're talking about here is _also_ computing `skips` for the same CSV. And writing a different `skips` list in the current CSV while leaving the old CSVs alone is exactly the use case described in the doc -- unless I'm wildly misunderstanding it.

Reply:
Honestly the doc isn't really clear (to me) for the use case we are trying to solve.  If you look at the example, X replaces X-2 and skips X-1.  To me, that meant we can't have a valid chain of X, X+1, X+2, X+3 (skip X+1) because the replaces in X+2 will be wrong as it would need to replace X, not X+1.

However, talking with the OLM team my understanding appears to be incorrect and yours appears to be correct.

One thing to be cautious about with skips, is that it will be ignored if we ever move (or are forced to move to) semantic versioning.  We might want to use skipRange instead as it would still be honored if we ever move to semantic versioning.

Reply:
Update on this.  After further discussion with the OLM team, it seems my interpretation for skips is correct.  It can only be used to skip n-1 and can not be used to define a list of previous versions to skip.  This is the same for skipRange.  There currently isn't a way to have the latest catalog define a list of previous versions to skip.

Reply:
I concur, with one quibble:

> n-1

It can be any number of contiguous versions from n-1 back. (If it's just n-1, it's technically unnecessary, redundant with `replaces` pointing at n-2.)

So we should set `skipRange` to cover the skipped version(s), as that'll get us happy with opm+semver. We could just always set it to all the known skips, because that's super easy to compute, and doesn't do any harm even when some values would be ignored. But at a minimum we should set it to cover the contiguous range up to n-1 iff n-1 is itself a skip.

A handy detail: `skipRange` accepts semver range semantics, so we could set it as (pseudocode) `skipRange = join(' || ', skiplist)`. That is, we don't need to compute inequalities; we can just enumerate the specific versions with logical OR.

Reply:
I've never understood the documentation of `skips` as all the versions to skip but to skip versions between two certain versions... a list instead of a `skipRange`. Exactly as the last message from @rrati

Reply:
@2uasimojo I don't think that is correct.  We can only set skips for csv after the one being skipped.  It's not possible to have the latest csv define a list of versions to skip.  The latest CSV can not define n-2 and n-1 skips if n-2 csv is already in the bundle and n-1 does not skip it.  Per the conversion:

We want the latest bundle in a catalog to determine which previous versions need to be skipped.  That way when we install the catalog fresh (as part of recovery) it will start at X-n and upgrade through all the operator versions except the ones listed in the skip of the latest bundle.

Yeah it's not possible today. We have some designs about how altering this package level metadata could work, but nothing is implemented yet.

Am I misunderstanding you?

Reply:
If I have understood https://coreos.slack.com/archives/C3VS0LV41/p1605117420420500 correctly, the only way to currently address the scenario in which we push something after the one we have to skip is the workaround I proposed. Is correct @rrati @2uasimojo?